File: RetosOtherLanguages/RetoN°4/RetoN°4.cpp
#include <iostream>
#include <stack>

class Decode {
public:
    static std::string decode(const std::string& message) {
        std::stack<char> stack;

        for (char c : message) {
            if (c == ')') {
                std::string reversed;
                while (!stack.empty() && stack.top() != '(') {
                    reversed += stack.top();
                    stack.pop();
                }
                // Remove the '(' from stack
                stack.pop();
                // Push the reversed string back into stack
                for (char reversedChar : reversed) {
                    stack.push(reversedChar);
                }
            } else {
                stack.push(c);
            }
        }

        std::string result;
        while (!stack.empty()) {
            result = stack.top() + result;
            stack.pop();
        }

        return result;
    }
};

int main() {
    std::string a = Decode::decode("hola (odnum)");
    std::cout << a << std::endl; // Hola mundo

    std::string b = Decode::decode("(olleh) (dlrow)!");
    std::cout << b << std::endl; // Hello World!

    std::string c = Decode::decode("sa(u(cla)atn)s");
    std::cout << c << std::endl; // santaclaus

    std::string d = Decode::decode("((nta)(sa))");
    std::cout << d << std::endl; // Hello World!

    return 0;
}

File: RetosOtherLanguages/RetoN°4/RetoN°4.php
<?php

class Decode {
    public static function decode($message) {
        $stack = new SplStack();

        foreach (str_split($message) as $c) {
            if ($c == ')') {
                $reversed = '';
                while (!$stack->isEmpty() && $stack->top() != '(') {
                    $reversed .= $stack->pop();
                }
                // Remove the '(' from stack
                $stack->pop();
                // Push the reversed string back into stack
                foreach (str_split($reversed) as $reversedChar) {
                    $stack->push($reversedChar);
                }
            } else {
                $stack->push($c);
            }
        }

        $result = '';
        while (!$stack->isEmpty()) {
            $result .= $stack->pop();
        }

        return strrev($result);
    }
}

$a = Decode::decode("hola (odnum)");
echo $a . PHP_EOL; // Hola mundo

$b = Decode::decode("(olleh) (dlrow)!");
echo $b . PHP_EOL; // Hello World!

$c = Decode::decode("sa(u(cla)atn)s");
echo $c . PHP_EOL; // santaclaus

$d = Decode::decode("((nta)(sa))");
echo $d . PHP_EOL; // Hello World!

?>
